src/iam-token.js

```diff
@@ -0,0 +1,146 @@
+      apiKey = await this._getSecretData(secretName, secretKey, secretNamespace, kubeResourceMeta, namespace);
+    } else if (typeof apiKeyRef == 'object') {
+      let secretName = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.name');
+      let secretNamespace = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.namespace', namespace);
+      let secretKey = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.key');
+      apiKey = await this._getSecretData(secretName, secretKey, secretNamespace, kubeResourceMeta, namespace);
+    }
+    if (!apiKey) {
+      return Promise.reject('Failed to find valid apikey to authenticate against iam');
+    }
+
+    let res = await this._requestToken(iam, apiKey);
+    return res.access_token;
+  }
+
+
+  async _requestToken(iam, apiKey) {
+    let token;
+
+    const apiKeyHash = hash(apiKey, { algorithm: 'shake256' });
+    // Check if the cache has the token. If it does have it, copy it into token
+    // for further processing
+    if (objectPath.has(this.s3TokenCache, [apiKeyHash])) {
+      // fetch cached token
+      token = objectPath.get(this.s3TokenCache, [apiKeyHash]);
+    }
+
+    // if we are able to fetch the token from the cache and it is not a Promise,
+    // then we need to check the expiry. If it is not expiring in the next 120 s
+    // return the token
+    if (!(token instanceof Promise)) {
+      if (token !== undefined) {
+        const expires = objectPath.get(token, 'expiration', 0); // expiration: time since epoch in seconds
+        // re-use cached token as long as we are more than 2 minutes away from expiring.
+        if (Date.now() < (expires - 120) * 1000) {
+          return token;
+        }
+      } else {
+        // if we get to this point the token is either undefined or it is stale data.
+        // either ways we need to connect to iam auth and get a new token
+        // create an async function that will wait for _getToken to return a token.
+        // since we are not using await, the control will return immediately and the
+        // return value of the tokenFunction is going to be a promise.
+        // we are going to cache this promise into the s3TokenCache and
+        // wait for it.
+        let tokenFunction = async (iam, apiKeyHash, apiKey) => {
+          let ret = await this._getToken(iam, apiKeyHash, apiKey);
+          return ret;
+        };
+        token = tokenFunction(iam, apiKeyHash, apiKey);
+        objectPath.set(this.s3TokenCache, [apiKeyHash], token);
+      }
+    }
+
+    // at this point token is either data, or Promise (either created in the else case above
+    // or already created by a different event)
+    if (token instanceof Promise) {
+      // if the token is a Promise, wait for it to be completed. Once completed, get
+      // the data from the cache and return it.
+      try {
+        await token;
+        return objectPath.get(this.s3TokenCache, [apiKeyHash]);
```

Comment:
This isn't setting the cache value, it would just return the promise.
```suggestion
        // token is a promise, wait for it to return a value
        const foundToken = await token;
        // value successfully retrieved, cache it
        objectPath.set(this.s3TokenCache, [apiKeyHash], token);
        return foundToken;
```
This could be done in fewer lines, e.g. `token = await token;` rather than defining a separate variable, but this makes it more clear.

Reply:
@carrolp I am setting the cache value in the async _getToken function. I don't need to set it again here. https://github.com/razee-io/RemoteResourceS3/pull/284/files#diff-8859986facd1f8d3d9d843b915e6aefce1a0bbc29ff45c28484f646f8a388288R132

Reply:
@rgowrishankar Thanks.  Functionally that should work, but setting a global as a side effect of a 'get XYZ' function is bad practice, makes the code hard to read.  E.g. even after being told that I had to read line 92 where it's setting the cache value to the promise multiple times to understand why it sets cache there, but not here.

Reply:
if the Promise should set the cached value (replacing the promise itself), then a better place would be in the promise function rather than _getToken, which should then **not** return the value, e.g.:
```
let tokenFunction = async (iam, apiKeyHash, apiKey) => {
        let ret = await this._getToken(iam, apiKeyHash, apiKey);
        objectPath.set(this.s3TokenCache, [apiKeyHash], ret);
      };
```
But I would prefer the version of code where neither function has 'side effects'.